### ISSUE TITLE
[CP] Maintain checked casts when using js_util in non-production mode

### ISSUE THREAD

**Issue by rileyporter** [dart-lang/sdk#48287]
**commit(s) to merge**: 150b51e and follow-up 9fff435 to fix the status files to skip csp tests

**merge instructions**: cleanly cherry-picks to "Version 2.16.0-134.6.beta"

**What is the issue**: We added a generic type parameter `T` as the return type to `js_util` methods. Before, when the return type of the `js_util` methods used to be `dynamic`, users had to cast the values `as T` themselves. After the change to the generic return type, users do not have to add a cast and there is no place where the return value from JavaScript is ever checked to actually be `T`.

**What is the fix**: We added an implicit downcast into `js_util`, by making the call to JavaScript return a `dynamic` value which would insert an implicit `as T` downcast on the value returned by JavaScript. In production mode, for dart2js with `-O3` flag, the implicit downcast will be removed so it will be optimized and not add a check for every `js_util` call. However, for DDC and dart2js when run with `-O2` and lower, the check `as T` will be added to get back to the same level of checking as before when `js_util` had a `dynamic` return type. The cast is just being moved from user code into `js_util` in this case.

The actual code change is ~20 lines of code in `js_util`, the rest of the cherry-pick is added test coverage for this case.

**Why cherrypick**: The generic return type change in `js_util` landed in d17fd4b on October 18th, 2021. Users could add usage of `js_util` that doesn't have a cast `as T` to check the JavaScript value, which would break when this change lands.

**Risk**: Little

**Link to original  issue(s)**: https://github.com/dart-lang/sdk/issues/47832

/cc @mit-mit @whesse @athomas @vsmenon @devoncarew @sigmundch

**Comment by sigmundch** [comment]
Thanks so much for filing the issue.

We discussed this at the scrum and decided in the end that we'll not cherry pick this. Instead, we'll include a CHANGELOG entry to announce that this change is breaking in 2.17 (but restoring old behavior),

The reason we decided to not cherry-pick is that the regression (which made cast more permissive) was already released in 2.15 as well (I thought initially it had only made it to 2.16.0), as such, making the breaking fix in a .1 release will be a bit surprising for end users.
